google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/PostgreSQLStatementParser.java

```diff
@@ -267,4 +290,32 @@ private void appendIfNotNull(
       result.append(prefix).append(tag).append(suffix);
     }
   }
+
+  private boolean isReturning(String sql, int index) {
+    // RETURNING is a reserved keyword in PG, but requires a
+    // leading AS to be used as column label, to avoid ambiguity.
+    // We thus check for cases which do not have a leading AS.
+    // (https://www.postgresql.org/docs/current/sql-keywords-appendix.html)
+    return (index >= 1)
+        && (index + 10 <= sql.length())
+        && RETURNING_PATTERN.matcher(sql.substring(index - 1, index + 10)).matches()
+        && !((index >= 4)
+            && AS_RETURNING_PATTERN.matcher(sql.substring(index - 4, index + 10)).matches());
+  }
+
+  @InternalApi
+  @Override
+  boolean checkReturningClauseInternal(String rawSql) {
+    Preconditions.checkNotNull(rawSql);
+    int index = 0;
+    String sql = rawSql.replaceAll("\\s+", " ").toLowerCase();
+    while (index < sql.length()) {
+      if (isReturning(sql, index)) {
+        return true;
+      } else {
+        index = skip(sql, index, null);
+      }
+    }
```

Comment:
This is still a very expensive check, as we always iterate through the entire SQL string and repeatedly check for a regex match. Can we improve this a little by doing the following:
1. Do a pre-check to see if the (lower)-case string contains `"returning"` at all. If it does not, we know for sure that the regex will also never match. See https://gist.github.com/olavloite/a857ff860c4f12c17d51712bed57b9d3 for a small benchmark program. Some quick tests indicate that having such a pre-check improves performance with a factor 10x (400ms for 100,000 checks as opposed to 4,000ms on my local machine).
2. Do the while loop in the opposite order (i.e. start at `index=sql.length()`). If we do a pre-check and that check returns true, the probability that we will find a match is a lot bigger, as `returning` is not something that is very common in any given SQL string. And a `returning` clause will be somewhere at the end of the SQL string.